### PR TITLE
chore: added sync skeleton

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**

### CODE REVIEW

File: src/syncManager.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        try {
+          this.logger.debug(`tiles signature is set to: ${this.tilesConfig.sigIsNeeded.toString()}`);
+          const generator = tilesGenerator(batch);
+          let batchArray = [];
+
+          for (const tile of generator) {
+            const path = `${this.tilesConfig.path}/${layerId}/${tile.zoom}/${tile.x}/${tile.y}.${this.tilesConfig.format}`;
+
+            if (await isFileExists(path)) {
+              batchArray.push(this.signAndUpload(tile, path));
```
Comment (asafMasa) [inline src/syncManager.ts:72]: i think we need to clone the file - sign it, send it and delete it
you don't want  to sign the original files

File: src/syncManager.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            }
+            if (batchArray.length === this.tilesConfig.uploadBatchSize) {
+              await Promise.all(batchArray);
+              await this.tilesManager.updateTilesCount(layerId, batchArray.length);
+              batchArray = [];
+            }
+            // resolved left overs
+            await Promise.all(batchArray);
```
Comment (asafMasa) [inline src/syncManager.ts:80]: will happen for every file or i am mistaken?

File: src/syncManager.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+          }
+
+          await this.tilesManager.updateTilesCount(layerId, batchArray.length);
+
+          await this.queueClient.queueHandler.ack(jobId, taskId);
+          await this.finishJob(jobId, layerId);
```
Comment (asafMasa) [inline src/syncManager.ts:86]: you can't determine if the job finished successfully from the worker - it's a bug

File: src/syncManager.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    await this.tilesManager.uploadTile(tile, path);
+  }
+
+  private async finishJob(jobId: string, layerId: string, isSuccess = true, errorReason: string | undefined = undefined): Promise<void> {
+    this.logger.info(`Update Job status to success=${String(isSuccess)} jobId=${jobId}`);
+    const joUpdatePayload: IUpdateJobRequestPayload = {
+      status: isSuccess ? TaskStatus.COMPLETED : TaskStatus.FAILED,
+      reason: errorReason,
+    };
+    await this.queueClient.queueHandler.jobManagerClient.updateJob(jobId, joUpdatePayload);
```
Comment (asafMasa) [inline src/syncManager.ts:110]: i think the ni-fi should determine the job status - talk with me

File: src/tilesManager.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      this.logger.error(`tiles count update failed for layerId=${layerId} with error: ${error}`);
+    }
+  }
+
+  public async uploadTile(tile: ITile, path: string): Promise<void> {
+    try {
+      await this.gatewayClient.upload(path);
+    } catch (error) {
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      this.logger.error(`upload failed for tile: ${tile.zoom}/${tile.x}/${tile.y} with error: ${error}`);
```
Comment (asafMasa) [inline src/tilesManager.ts:38]: instead of ${tile.zoom}/${tile.x}/${tile.y} use ${path}

**Review by asafMasa — CHANGES_REQUESTED** [review]
Great job

File: config/custom-environment-variables.json [hunk 6]
```diff
@@ -38,5 +38,46 @@
...
+  },
+  "gateway": {
+    "url": "GATEWAY_URL"
+  },
+  "tiles": {
+    "path": "TILES_PATH",
+    "format": "TILES_FORMAT",
+    "uploadBatchSize": {
+      "__name": "TILEs_UPLOAD_BATCH_SIZE",
+      "__format": "boolean"
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply config/custom-environment-variables.json:56]: fixed

File: config/default.json [hunk 7]
```diff
@@ -7,12 +7,12 @@
   },
   "telemetry": {
     "logger": {
-      "level": "info",
+      "level": "debug",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply config/default.json:10]: fixed

File: config/default.json [hunk 8]
```diff
@@ -24,5 +24,36 @@
...
     }
+  },
+  "layerSpecBaseUrl": "http://localhost:8080",
+  "nifiBaseUrl": "http://localhost:8081",
+  "syncAttempts": 5,
+  "gateway": {
+    "url": "http://localhost:6969"
+  },
+  "tiles": {
+    "path": "/home/shlomiko/Desktop",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply config/default.json:35]: fixed

File: config/default.json [hunk 9]
```diff
@@ -24,5 +24,36 @@
...
+    "format": "png",
+    "uploadBatchSize": 100,
+    "sigIsNeeded": true
+  },
+  "crypto": {
+    "readFileEncoding": "ascii",
+    "pem": "tests/mocks/files/invalidKey.pem",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply config/default.json:42]: fixed

File: openapi3.yaml [hunk 10]
```diff
@@ -1,51 +1,50 @@
 openapi: 3.0.1
 info:
-  title: service-name
+  title: Sync Worker
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply openapi3.yaml:3]: removed

File: package.json [hunk 11]
```diff
@@ -5,7 +5,7 @@
   "main": "./src/index.ts",
   "scripts": {
     "test:unit": "jest --config=./tests/configurations/unit/jest.config.js",
-    "test:integration": "jest --config=./tests/configurations/integration/jest.config.js",
+    "test:integration": "jest --passWithNoTests --config=./tests/configurations/integration/jest.config.js",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply package.json:8]: removed

File: package.json [hunk 12]
```diff
@@ -42,17 +42,22 @@
...
     "@map-colonies/js-logger": "^0.0.3",
+    "@map-colonies/mc-priority-queue": "^2.0.3",
+    "@map-colonies/mc-utils": "^1.1.2",
     "@map-colonies/openapi-express-viewer": "^2.0.1",
     "@map-colonies/telemetry": "3.0.0",
     "@opentelemetry/api": "1.0.1",
     "@opentelemetry/api-metrics": "0.23.0",
     "@opentelemetry/instrumentation-express": "0.23.0",
     "@opentelemetry/instrumentation-http": "0.23.0",
+    "axios": "^0.21.1",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply package.json:53]: it is needed for unit-test axios mocks.

File: src/clients/services/layerSpecClient.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import { Services } from '../../common/constants';
+
+interface ITilesCountUpdateRequest {
+  tilesBatchCount: number;
+}
+
+@singleton()
+export class LayerSpecClient extends HttpClient {
+  public constructor(@inject(Services.LOGGER) logger: Logger, @inject(Services.CONFIG) private readonly config: IConfig) {
+    super(logger, '', 'layer-spec', config.get<IHttpRetryConfig>('httpRetry'));
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/layerSpecClient.ts:14]: fixed

File: src/clients/services/gatewayClient.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+import FormData from 'form-data';
+import { Logger } from '@map-colonies/js-logger';
+import { HttpClient, IHttpRetryConfig } from '@map-colonies/mc-utils';
+import { IConfig } from '../../common/interfaces';
+import { Services } from '../../common/constants';
+
+@singleton()
+export class GatewayClient extends HttpClient {
+  public constructor(@inject(Services.LOGGER) logger: Logger, @inject(Services.CONFIG) private readonly config: IConfig) {
+    super(logger, config.get<string>('gateway.url'), 'upload', config.get<IHttpRetryConfig>('httpRetry'));
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/gatewayClient.ts:12]: fixed

File: src/clients/services/layerSpecClient.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    this.axiosOptions.baseURL = config.get<string>('layerSpecBaseUrl');
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/layerSpecClient.ts:15]: fixed

File: src/clients/services/layerSpecClient.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  public async updateTilesCount(layerId: string, tilesCount: number): Promise<void> {
+    const updateRequest: ITilesCountUpdateRequest = {
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/layerSpecClient.ts:19]: fixed

File: src/clients/services/nifiClient.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+export interface OnSuccessUpdateRequest {
+  layerId: string;
+  jobId: string;
+}
+
+@singleton()
+export class NifiClient extends HttpClient {
+  public constructor(@inject(Services.LOGGER) logger: Logger, @inject(Services.CONFIG) private readonly config: IConfig) {
+    super(logger, '', 'nifi', config.get<IHttpRetryConfig>('httpRetry'));
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/nifiClient.ts:15]: fixed

File: src/clients/services/nifiClient.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    this.axiosOptions.baseURL = config.get<string>('nifiBaseUrl');
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/nifiClient.ts:16]: fixed

File: src/clients/services/nifiClient.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  public async notifyNifiOnSuccess(jobId: string, layerId: string): Promise<void> {
+    const body: OnSuccessUpdateRequest = {
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/nifiClient.ts:20]: fixed

File: src/syncManager.ts [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    @inject(Services.LOGGER) private readonly logger: Logger,
+    @inject(Services.CONFIG) private readonly config: IConfig,
+    @inject(Services.TILES_CONFIG) private readonly tilesConfig: ITilesConfig,
+    @inject(Services.CRYPTO_CONFIG) private readonly cryptoConfig: ICryptoConfig,
+    private readonly queueClient: QueueClient,
+    private readonly tilesManager: TilesManager,
+    private readonly cryptoManager: CryptoManager,
+    private readonly nifiClient: NifiClient
+  ) {
+    this.logger = logger;
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:47]: removed

File: src/syncManager.ts [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+      const params = task.parameters as IParameters;
+      const jobId = task.jobId;
+      const taskId = task.id;
+      const batch = params.batch;
+      const attempts = task.attempts;
+      const layerId = `${params.resourceId}-${params.resourceVersion}`;
+
+      if (attempts <= this.syncAttempts) {
+        try {
+          this.logger.debug(`tiles signature is set to: ${this.tilesConfig.sigIsNeeded.toString()}`);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:64]: fixed

File: src/syncManager.ts [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+  ) {
+    this.logger = logger;
+    this.syncAttempts = this.config.get<number>('syncAttempts');
+  }
+
+  public async runSync(): Promise<void> {
+    const task = await this.queueClient.queueHandler.waitForTask();
+
+    if (task) {
+      const params = task.parameters as IParameters;
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:55]: fixed

File: src/syncManager.ts [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+      const layerId = `${params.resourceId}-${params.resourceVersion}`;
+
+      if (attempts <= this.syncAttempts) {
+        try {
+          this.logger.debug(`tiles signature is set to: ${this.tilesConfig.sigIsNeeded.toString()}`);
+          const generator = tilesGenerator(batch);
+          let batchArray = [];
+
+          for (const tile of generator) {
+            const path = `${this.tilesConfig.path}/${layerId}/${tile.zoom}/${tile.x}/${tile.y}.${this.tilesConfig.format}`;
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:69]: done

File: src/syncManager.ts (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:72]: fixed

File: src/tilesManager.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/tilesManager.ts:38]: fixed

**Review by asafMasa — CHANGES_REQUESTED** [review]
"name": "service-name" - find in all project

File: src/clients/services/gatewayClient.ts [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+@singleton()
+export class GatewayClient extends HttpClient {
+  public constructor(@inject(Services.LOGGER) logger: Logger, @inject(Services.CONFIG) private readonly config: IConfig) {
+    super(logger, config.get<string>('gateway.url'), 'GatewayClient', config.get<IHttpRetryConfig>('httpRetry'));
+  }
+
+  public async upload(buffer: Buffer): Promise<void> {
+    const formData = new FormData();
+    formData.append('photo', buffer, { filename: 'filename.png' });
```
Comment (asafMasa) [inline src/clients/services/gatewayClient.ts:16]: Add todo - fix in integration

File: src/cryptoManager.ts [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+@singleton()
+export class CryptoManager {
+  public constructor(@inject(Services.LOGGER) private readonly logger: Logger) {
+    this.logger = logger;
+  }
+
+  public async generateSignedFile(keyFilePath: string, filePath: string, buffer: Buffer): Promise<Buffer> {
+    try {
+      const hash = await this.computeHash(keyFilePath);
```
Comment (asafMasa) [inline src/cryptoManager.ts:20]: check what you sign and with who

File: src/syncManager.ts [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+  private async finishJob(jobId: string, layerId: string, isSuccess = true, errorReason: string | undefined = undefined): Promise<void> {
+    this.logger.info(`Update Job status to success=${String(isSuccess)} jobId=${jobId}`);
+    const joUpdatePayload: IUpdateJobRequestPayload = {
+      status: isSuccess ? TaskStatus.COMPLETED : TaskStatus.FAILED,
+      reason: errorReason,
+    };
+    await this.queueClient.queueHandler.jobManagerClient.updateJob(jobId, joUpdatePayload);
+    if (isSuccess) {
+      this.logger.info(`Update Nifi on success for jobId=${jobId}, layerId=${layerId}`);
+      await this.nifiClient.notifyNifiOnSuccess(jobId, layerId);
```
Comment (asafMasa) [inline src/syncManager.ts:115]: also on failed state

**Review by asafMasa — CHANGES_REQUESTED** [review]
remove from workflow - OpenAPI Lint Checks

File: src/syncManager.ts (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:81]: fixed

File: src/syncManager.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:87]: fixed

File: src/syncManager.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:112]: fixed

File: src/clients/services/gatewayClient.ts (hunk 24 above)
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/gatewayClient.ts:16]: added

File: src/syncManager.ts (hunk 26 above)
Comment (CL-SHLOMIKONCHA) [reply src/syncManager.ts:115]: fixed

File: src/cryptoManager.ts (hunk 25 above)
Comment (CL-SHLOMIKONCHA) [reply src/cryptoManager.ts]: fixed

**Review by yossiz16 — COMMENTED** [review]
helm approved

File: src/cryptoManager.ts [hunk 27]
```diff
@@ -30,26 +36,36 @@ export class CryptoManager {
     }
   }
 
-  private async computeHash(keyFilePath: string): Promise<string> {
-    this.logger.debug('computing hash key');
-    const secret = await fsp.readFile(keyFilePath, { encoding: 'binary' });
-    const hash = crypto.createHash('sha256');
+  private async computeHash(filePath: string): Promise<Buffer> {
+    const secret = await fsp.readFile(filePath, { encoding: 'binary' });
+    const hash = crypto.createHash('sha512');
```
Comment (asafMasa) [inline src/cryptoManager.ts:41]: move to config (no need for environment variable)

File: src/cryptoManager.ts [hunk 28]
```diff
@@ -30,26 +36,36 @@ export class CryptoManager {
...
     return hashKey;
   }
 
-  private encryptHash(fileHash: string): IEncryptedHash {
+  private encryptHash(fileHash: Buffer): IEncryptedHash {
     const ivSize = 16;
-    this.logger.debug('encrypting hash');
     const iv = Buffer.allocUnsafe(ivSize);
-    const cipher = crypto.createCipheriv('aes-256-cfb', Buffer.from(fileHash, 'base64'), iv);
+    const cipher = crypto.createCipheriv('aes-256-cfb', this.key, iv);
```
Comment (asafMasa) [inline src/cryptoManager.ts:50]: move to config (no need for environment variable)

File: src/clients/services/gatewayClient.ts [hunk 29]
```diff
@@ -13,6 +13,7 @@ export class GatewayClient extends HttpClient {
 
   public async upload(buffer: Buffer): Promise<void> {
     const formData = new FormData();
+    // TODO: fix in integration
```
Comment (asafMasa) [inline src/clients/services/gatewayClient.ts:16]: you didn't resolved my remark of "i dont think you need to specify the baseUrl"

File: src/cryptoManager.ts [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+    return encryptedHash;
+  }
+
+  private readKeyFile(filePath: string): string {
+    try {
+      const key = readFileSync(filePath, { encoding: 'utf8' });
+      return key;
+    } catch (error) {
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      this.logger.fatal(`Could not read key from provided file path: ${filePath}, error: ${error}`);
```
Comment (asafMasa) [inline src/cryptoManager.ts:67]: you didn't resolve my comment of "make sure error is realy printed instead of "object" word"

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍

File: src/cryptoManager.ts (hunk 27 above)
Comment (CL-SHLOMIKONCHA) [reply src/cryptoManager.ts:41]: fixed

File: src/cryptoManager.ts (hunk 28 above)
Comment (CL-SHLOMIKONCHA) [reply src/cryptoManager.ts:50]: fixed

File: src/clients/services/gatewayClient.ts (hunk 29 above)
Comment (CL-SHLOMIKONCHA) [reply src/clients/services/gatewayClient.ts:16]: fixed

File: src/cryptoManager.ts (hunk 30 above)
Comment (CL-SHLOMIKONCHA) [reply src/cryptoManager.ts:67]: done.

File: config/custom-environment-variables.json [hunk 31]
```diff
@@ -38,5 +38,46 @@
...
+    "sigIsNeeded": {
+      "__name": "TILES_SIG_IS_NEEDED",
+      "__format": "boolean"
+    }
+  },
+  "crypto": {
+    "readFileEncoding": "CRYPTO_READ_FILE_ENCODING",
+    "pem": "CRYPTO_PEM",
+    "algoritm": "CRYPTO_ALOGRITM",
+    "sigEncoding": "CRYPTO_SIG_ENCODING"
```
Comment (asafMasa) [inline config/custom-environment-variables.json:67]: remove?

**Review by asafMasa — APPROVED** [review]

**Pull request merged**